contracts/VolOracle.sol

```diff
@@ -59,24 +61,45 @@ contract VolOracle {
         // set the tickSquareCumulative to 0 during initialization
         oracleState.observations[0] = VolObservation(blockTimestamp, tickCumulative, 0);
         // TODO: should we use current blocktimestamp or the last observation timestamp from uni here
+
         oracleState.lastBlockTimestamp = block.timestamp;
         oracleState.lastObservationIndex = observationIndex;
         oracleState.observationIndex = 0;
     }
 
     // returns the start and end indexes for filling intermediate values
     function fetchIntermediateIndexes(address _pool) public view returns (uint256 startIndex, uint256 endIndex) {
-        (, , uint16 latestPoolObservationIndex, uint16 poolCardinality, , , ) = IUniswapV3Pool(_pool).slot0();
-        uint256 lastObservationIndex = oracleStates[_pool].lastObservationIndex;
-        // TODO: we are assuming the oracle won't be down for a long time (>7 days here)
-        // both inclusive
-        if (latestPoolObservationIndex > lastObservationIndex)
-            return (lastObservationIndex + 1, latestPoolObservationIndex);
-        return (lastObservationIndex + 1, latestPoolObservationIndex + poolCardinality);
+        IUniswapV3Pool uniPool = IUniswapV3Pool(_pool);
+        (, , uint16 latestPoolObservationIndex, uint16 poolCardinality, , , ) = uniPool.slot0();
+
+        uint256 oldestObservationIndex = (latestPoolObservationIndex + 1) % poolCardinality;
+
+        (uint32 oldestObservationTs, , , bool initialized) = uniPool.observations(oldestObservationIndex);
```

Comment:
add this check as well: https://github.com/Uniswap/v3-periphery/blob/main/contracts/libraries/OracleLibrary.sol#L83